pkg/ingester/retention.go

```diff
@@ -39,183 +66,363 @@ func defaultRetentionPolicy() retentionPolicy {
+		return 0, 0, false
 	}
-	// Get all block across all the tenants. Any block
-	// produced or imported during the procedure is ignored.
-	blocks, err := e.localBlocks(e.dbConfig.DataPath)
+
+	tenantIDs, err := dc.blockManager.GetTenantIDs(ctx)
 	if err != nil {
-		return err
+		level.Error(dc.logger).Log(
+			"msg", "failed run high disk cleanup, could not read tenant ids",
+			"err", err,
+		)
+		return 0, 0, true
 	}
 
-	for volumeStatsCurrent.HighDiskUtilization && len(blocks) > 0 && ctx.Err() == nil {
-		// When disk utilization is not lower since the last loop, we end the
-		// cleanup there to avoid deleting all blocks when disk usage reporting
-		// is delayed.
-		if volumeStatsPrev != nil && volumeStatsPrev.BytesAvailable >= volumeStatsCurrent.BytesAvailable {
-			level.Warn(e.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
+	blocks := make([]*tenantBlock, 0)
+	for _, tenantID := range tenantIDs {
+		tenantBlocks, err := dc.blockManager.GetBlocksForTenant(ctx, tenantID)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to get blocks for tenant",
+				"tenantID", tenantID,
+				"err", err,
+			)
+
+			// Keep trying to read blocks from other tenants.
+			continue
+		}
+
+		blocks = append(blocks, tenantBlocks...)
+	}
+
+	// Sort by uploaded, then age (oldest first).
+	sort.Sort(blocksByUploadAndAge(blocks))
+
+	prevVolumeStats := &diskutil.VolumeStats{}
+	filesDeleted := 0
+	for _, block := range blocks {
+		// Delete a block.
+		err = dc.blockManager.DeleteBlock(ctx, block)
+		switch {
+		case os.IsNotExist(err):
+			level.Warn(dc.logger).Log(
+				"msg", "failed to delete uploaded block, does not exist",
+				"err", err,
+				"path", block.Path,
+			)
+			return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
+		case err != nil:
+			level.Error(dc.logger).Log(
+				"msg", "failed run high disk cleanup, could not delete block",
+				"path", block.Path,
+				"err", err,
+			)
+			return filesDeleted, int(volumeStats.BytesAvailable - prevVolumeStats.BytesAvailable), true
+		default:
+			filesDeleted++
+		}
+
+		// Recheck volume stats.
+		prevVolumeStats = volumeStats
+		volumeStats, err = dc.volumeChecker.HasHighDiskUtilization(dc.config.DataPath)
+		if err != nil {
+			level.Error(dc.logger).Log(
+				"msg", "failed to check disk utilization",
+				"err", err,
+			)
 			break
 		}
-		// Delete the oldest block.
-		var b *tenantBlock
-		b, blocks = blocks[0], blocks[1:]
-		level.Warn(e.logger).Log("msg", "disk utilization is high, deleting the oldest block", "path", b.path)
-		if err = e.deleteBlock(b); err != nil {
-			return err
+
+		// Check if deletion should stop.
+		if !volumeStats.HighDiskUtilization || prevVolumeStats.BytesAvailable >= volumeStats.BytesAvailable {
+			level.Warn(dc.logger).Log("msg", "disk utilization is not lowered by deletion of a block, pausing until next cycle")
```

Comment:
The log message here reads a bit strange, according to the condition - the utilization was lowered or am I reading it wrong?

Reply:
The logic for this case was all messed up. Basically, there are three conditions (besides errors) which cause use to break the loop.

1. No more blocks to delete
2. Disk utilization is no longer reported high
3. Available bytes has not changed after deleting a block -- this can happen if `VolumeChecker` is delayed when reporting the current disk stats, in which case we want to be conservative and stop deleting blocks

In https://github.com/grafana/pyroscope/pull/2816/commits/e61b35a2bd2f8f98c8cba4658506c2603e54175c I did some clean up to break out of the loop in appropriate cases.

Reply:
Much easier to read now!